File: analysis/Python/Trend.py
from enum import Enum

class Direction(Enum):
    UP, DOWN = range(2)

def weightingInterval(n):
    return 1 / (n + 1)

def weightedAverage(numbers):
    interval = weightingInterval(len(numbers))
    average = 0
    current_interval = 1 - interval
    for num in numbers:
        average += current_interval * num
        # print(f'Current Interval: {current_interval}')
        # print(f'Current Average: {average}')
        current_interval -= interval
    # print(f'Weighted Average: {average}')
    average = average / (len(numbers) / 2.0)
    # print(f'New Average: {average}')
    return average

def percentError(a, b):
    return ((a - b) / b) * 100

def fluctuationFirstToRest(list):
    if len(list) == 1:
        return 0
    return percentError(list[0], weightedAverage(list[1:]))

def fluctuation(list):
    fluc = fluctuationFirstToRest(list)
    return fluctuationDirection(fluc)

def fluctuationDirection(fluc):
    # print(f'Passed in fluctuation: {fluc}')
    if fluc > 15:
        return Direction.UP
    elif fluc < -15:
        return Direction.DOWN
    else:
        return None

File: analysis/Haskell/src/Trend.hs
module Trend
(
    Direction(..),
    fluctuation,
    percentError
)
where
    import Data.Function;

    data Direction = Up | Down deriving (Show);

    -- premise is that starting values get more weighting than ending values.
    -- the way we accomplish this is by distributing weightings from the range [1..0] evenly across the list.
    -- we can accomplish this by getting the interval: (1 / (n + 1)), and creating the weights by subtracting this from 1
    weightedAverage :: [Double] -> Double
    weightedAverage ls = (/ fromIntegral len) $ sum $ zipWith (*) ls $ weightingList len
      where
        len = length ls

    -- given a list, yields the list of weightings to use.
    -- did it this way so that the first one would not be 1.0
    weightingList :: Int -> [Double]
    weightingList len = take len $ drop 1 [1, 1 - interval ..]
      where
        interval = 1 / fromIntegral (len + 1)

    percentError :: Double -> Double -> Double
    percentError a b = ((a - b) / b) * 100

    fluctuationFirstToRest :: [Double] -> Double
    fluctuationFirstToRest (x:xs) = percentError x (weightedAverage xs)

    fluctuation :: [Double] -> Maybe Direction
    fluctuation (x:xs) = fluctuationDirection $ fluctuationFirstToRest (x:xs)

    fluctuationDirection :: Double -> Maybe Direction
    fluctuationDirection f
        | f > 15 = Just Up
        | f < -15 = Just Down
        | otherwise = Nothing

    {-

    (1 - (1 * 0.2))x +
    (1 - (2 * 0.2))x-1 +
    (1 - (3 * 0.2))x-2 +
    (1 - (4 * 0.2))x-3 

    0.8 + 0.6 + 0.4 + 0.2 = 2

    1 - 1 * 0.1 +
    1 - 2 * 0.1 +
    1 - 3 * 0.1 +
    1 - 4 * 0.1 +
    1 - 5 * 0.1 +
    1 - 6 * 0.1 +
    1 - 7 * 0.1 +
    1 - 8 * 0.1 +
    1 - 9 * 0.1

    0.9 + 0.8 + 0.7 + 0.6 + 0.5 + 0.4 + 0.3 + 0.2 + 0.1 = 4.5

    -}
